cms/djangoapps/contentstore/rest_api/v1/views/course_index.py

```diff
@@ -0,0 +1,103 @@
+
+    @apidocs.schema(
+        parameters=[
+            apidocs.string_parameter("course_id", apidocs.ParameterLocation.PATH, description="Course ID"),
+            apidocs.string_parameter(
+                "show",
+                apidocs.ParameterLocation.QUERY,
+                description="Query param to set initial state which fully expanded to see the item",
+            )],
+        responses={
+            200: CourseIndexSerializer,
+            401: "The requester is not authenticated.",
+            403: "The requester cannot access the specified course.",
+            404: "The requested course does not exist.",
+        },
+    )
+    @verify_course_exists()
+    def get(self, request: Request, course_id: str):
+        """
+        Get an object containing course index for outline.
+
+        **Example Request**
+
+            GET /api/contentstore/v1/course_index/{course_id}?show=block-v1:edx+101+y+type@course+block@course
+
+        **Response Values**
+
+        If the request is successful, an HTTP 200 "OK" response is returned.
+
+        The HTTP 200 response contains a single dict that contains keys that
+        are the course's outline.
+
+        **Example Response**
+
+        ```json
+        {
+            "course_release_date": "Set Date",
+            "course_structure": {},
+            "deprecated_blocks_info": {
+                "deprecated_enabled_block_types": [],
+                "blocks": [],
+                "advance_settings_url": "/settings/advanced/course-v1:edx+101+y76"
+            },
+            "discussions_incontext_feedback_url": "",
+            "discussions_incontext_learnmore_url": "",
+            "initial_state": {
+                "expanded_locators": [
+                "block-v1:edx+101+y76+type@chapter+block@03de0adc9d1c4cc097062d80eb04abf6",
+                "block-v1:edx+101+y76+type@sequential+block@8a85e287e30a47e98d8c1f37f74a6a9d"
+                ],
+                "locator_to_show": "block-v1:edx+101+y76+type@chapter+block@03de0adc9d1c4cc097062d80eb04abf6"
+            },
+            "initial_user_clipboard": {
+                "content": null,
+                "source_usage_key": "",
+                "source_context_title": "",
+                "source_edit_url": ""
+            },
+            "language_code": "en",
+            "lms_link": "//localhost:18000/courses/course-v1:edx+101+y76/jump_to/block-v1:edx+101+y76",
+            "mfe_proctored_exam_settings_url": "",
+            "notification_dismiss_url": "/course_notifications/course-v1:edx+101+y76/2",
+            "proctoring_errors": [],
+            "reindex_link": "/course/course-v1:edx+101+y76/search_reindex",
+            "rerun_notification_id": 2
+        }
+        ```
+        """
+
+        course_key = CourseKey.from_string(course_id)
+        if not has_studio_read_access(request.user, course_key):
+            self.permission_denied(request)
+
+        with modulestore().bulk_operations(course_key):
```

Comment:
Hi, Jesper here from TNL. Just a side comment, not a full review: I've been working with contentstore APIs a lot recently and we're trying to make API views separate from business logic now. [Here's](https://github.com/openedx/edx-platform/blob/master/cms/djangoapps/contentstore/rest_api/v0/views/assets.py) an example. The reason is that we now have different APIs using the same logic - django API, REST API for MFEs, public authoring API. I basically introduced something similar to a service layer via files that are called e.g. `asset_storage_handler.py`.

Concretely, I would like it if you could extract this logic that calls modulestore directly to another file. The view here would be responsible for example for permission checks like above, and adding a serializer of course, but it would be agnostic to where the data is coming from - whether modulestore or not - and how it is handled. So basically lines 94-100 should be extracted in my opinion.

(a good rule of thumb for this: if logic does API-management stuff like serializing and validating data and checking for permissions, this should live in the view, and if the logic calls modulestore or retrieves, maps, formats, changes the data, it should not live in the view file.)

Reply:
@connorhaugh @jesperhodge Thanks for the suggestions. 

>and how requirements around performance (or lack thereof because this is studio ) informed your choices here.

@connorhaugh Actually, this work was passed on to us by someone else so we just rebased it and added some tests. But, I'll surely go through `learning_sequences` and see if we can update this PR to improve performance.

Reply:
@connorhaugh I did go through the learning sequences module and currently it seems to be designed for LMS i.e. it doesn't store any information about changes to course outline in draft mode. It will require some major changes to support draft changes.

So I propose to have this PR merged (after full review and applying @jesperhodge's suggestions) and once we are done with the building out the course outline in course-authoring MFE we should have a better idea of the data required. I'll schedule a ticket to migrate to learning sequences at that point.

Reply:
@jesperhodge Following your advice, I have moved usage of modulestore out of the new API view file into the `get_course_index_context` util.

Reply:
Thanks, that looks good! :) I'll resolve this conversation.

Just for reference so that you know what I was talking about, there's an [approved ADR](https://github.com/openedx/edx-platform/pull/33815) about this I'll merge soon.